azerothcore/azerothcore-wotlk
# feat(Core/Hooks): Add parameter to detect XP origin for OnGiveXP hook.

## Changes Proposed:
-  Add an enum parameter to detect where XP originated from.

## Issues Addressed:

- None

## SOURCE:


## Tests Performed:

- Tested using breakpoints to ensure it runs properly.


## How to Test the Changes:


1. Set breakpoint on line 159 of PlayerScript.cpp
2. Gain XP
3. Observe the local values, namely xpSource and ensure it matches the enum values.
### Values
```cpp
enum PlayerXPSource
{
    XPSOURCE_KILL = 0,
    XPSOURCE_QUEST = 1,
    XPSOURCE_QUEST_DF = 2,
    XPSOURCE_EXPLORE = 3,
    XPSOURCE_BATTLEGROUND = 4
};
```

## Known Issues and TODO List:


- This is a breaking change for modules.
- Modules will have to do extra validation since I moved the hook outside of Player::GiveXP.
### Missing Validation:
```cpp
    if (xp < 1)
    {
        return;
    }

    if (!IsAlive() && !GetBattlegroundId() && !isLFGReward)
    {
        return;
    }

    if (HasPlayerFlag(PLAYER_FLAGS_NO_XP_GAIN))
    {
        return;
    }

    if (victim && victim->GetTypeId() == TYPEID_UNIT && !victim->ToCreature()->hasLootRecipient())
    {
        return;
    }
```


## How to Test AzerothCore PRs
 
When a PR is ready to be tested, it will be marked as **[WAITING TO BE TESTED]**.

You can help by testing PRs and writing your feedback here on the PR's page on GitHub. Follow the instructions here:

http://www.azerothcore.org/wiki/How-to-test-a-PR

**REMEMBER**: when testing a PR that changes something **generic** (i.e. a part of code that handles more than one specific thing), the tester should not only check that the PR does its job (e.g. fixing spell XXX) but **especially** check that the PR does not cause any regression (i.e. introducing new bugs).

**For example**: if a PR fixes spell X by changing a part of code that handles spells X, Y, and Z, we should not only test X, but **we should test Y and Z as well**.
